### DISCUSSION TITLE
Que padrões vocês adotam quando criam um projeto do zero?

### DISCUSSION THREAD

**Discussion by renedocket** [frontendbr/forum#1885] [Perguntas e Respostas] [answered]
Olá, pessoal! Tudo beleza?

Eu vejo que muitas vagas de emprego pedem pro candidato já ter feito algum projeto do zero, usando algum tipo de padrão visando escalabilidade, manutenibilidade, formas de code review e afins.
Visando um cenário em time, no caso se você fosse a pessoa responsável pela arquitetura do front-end, como vocês começariam o projeto, que tecnologias usariam e por que?
E que tipo de padrões seriam adotados, sejam em forma de escrever, commitar e por último não menos importante que ferramentas usariam como auxilo?

Obs: Eu me baseei em vagas front-end que pediam coisas relacionadas ao tópico, não sei de fato na prática é assim, eu gostaria de saber a opinião de vocês sobre esses pontos, e se eles se fazem validos no dia a dia nas empresas.

**Comment by ninetails** [comment, marked as answer]
Opa, separaria a questão em dois assuntos distintos. O primeiro é que o ideal é vc conhecer o máximo de ferramentas existentes para assegurar qualidade de código (pelo menos saber da existência mas se tiver brincado com elas, melhor). A lista é extensa. Mas envolvendo frontend, tem padrão de comunidade pra tudo quase rs. Só pra citar alguns e finalidades:

- Padrões de escrita de JS como [Standard JS](https://standardjs.com/) e [airbnb](https://github.com/airbnb/javascript), ou gerais como do [Google](https://google.github.io/styleguide/);
- Linters para validar escrita de código tanto de JS/TS (eslint, jshint, tslint....) quanto de CSS (stylelint) quanto pra arquivos gerais (Prettier);
- TypeScript pra adicionar tipagem estática no projeto;
- Validadores de front [W3C](https://validator.w3.org/) para assegurar _web standards_;
- Validadores gerais mas que também vêem performance, como [Lighthouse](https://developers.google.com/web/tools/lighthouse) e [GTMetrix](https://gtmetrix.com/);
- Validadores para acessibilidade e WAI-ARIA como o axe;
- Linter pra mensagem de commits como o [commitlint](https://github.com/conventional-changelog/commitlint);
- Padrões para mensagens de commit como o [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/);
- Padrões para versionamento como o [Semantic Versioning](https://semver.org/);
- Fluxos de desenvolvimento usando Git, como [Git flow](https://danielkummer.github.io/git-flow-cheatsheet/), [Github Flow](https://guides.github.com/introduction/flow/) que daí inclui nomenclaturas de branches etc;
- Frameworks podem ter alguns padrões associados também...
- Não esquecer de testes unitários, de integração e e2e;
- Ferramentas também de análise estática de código como [Code Climate](https://codeclimate.com/) e [SonarQube](https://www.sonarqube.org/);
- ...e isso tudo integrado em alguma ferramenta de <abbr title="Continuous Integration / Continuous Delivery">CI / CD</abbr>

E com certeza devo ter esquecido de muita coisa por aí. A finalidade é de fazer um _hot dog_ e ir incluindo tudo o que faça com que seu projeto tenha maior qualidade, seja menos suscetível a bugs, tenha uma escrita uniforme entre todos os desenvolvedores facilitando a leitura do código...

Enfim o segundo ponto: padrões. Listei alguns acima, mas quando se trata de uma empresa, é necessário seguir os padrões definidos formal ou informalmente dentro dela. Como disse, pra assegurar que seja algo uniforme entre todos os devs. Pode ser algum padrão decidido entre os integrantes quanto o padrão que já existir quando adotado pelo primeiro dev do projeto e não foi mudado/questionado até então. Dado isso, ideal é evitar adicionar padrões conflitantes, por exemplo: o projeto utilizar StandardJS e alguém querer mudar para Airbnb, pois vai quebrar tal uniformidade de escrita de código e dificultar a manutenção/leitura do projeto.

Dado isso, também é sempre bom você ter os teus padrões já bem definidos, pois algum dia você vai precisar criar algum projeto. Seja pessoal, seja em alguma empresa. Quando a empresa não tem padrão definido, você irá precisar definir um... e como nós desenvolvedores somos bem egocêntricos, claro que escolhemos a que nos mais agrada rs.

<sup><sub>Desculpe pelo texto confuso pois fui escrevendo o que vinha em meus pensamentos.</sub></sup>

Um adendo: estude mas falo isso de maneira bem pessoal. Se a vaga for de nível junior ou estagiário é bom saber mas não é obrigação saber. Não deveria ser requisito de nenhuma vaga do nível.
